vm/systemSmartContracts/delegationManager.go

```diff
@@ -273,34 +288,56 @@ func (d *delegationManager) makeNewContractFromValidatorData(args *vmcommon.Cont
 		d.eei.AddReturnMessage(err.Error())
 		return vmcommon.OutOfGas
 	}
-	if len(args.Arguments) != 2 {
-		d.eei.AddReturnMessage("invalid number of arguments")
+	if core.IsSmartContractAddress(args.CallerAddr) {
+		d.eei.AddReturnMessage("cannot change from validator to delegation contract for a smart contract")
 		return vmcommon.UserError
 	}
 
-	arguments := append([][]byte{args.CallerAddr}, args.Arguments...)
-
-	return d.deployNewContract(args, false, initFromValidatorData, arguments)
+	return vmcommon.Ok
 }
 
 func (d *delegationManager) mergeValidatorDataToContract(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	if !d.flagValidatorToDelegation.IsSet() {
-		d.eei.AddReturnMessage("invalid function to call")
+	returnCode := d.checkValidatorToDelegationInput(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	if len(args.Arguments) != 1 {
+		d.eei.AddReturnMessage("invalid number of arguments")
 		return vmcommon.UserError
 	}
-	if args.CallValue.Cmp(zero) != 0 {
-		d.eei.AddReturnMessage("callValue must be 0")
+	lenAddress := len(args.CallerAddr)
+	scAddress := args.Arguments[0]
+	if len(scAddress) != lenAddress {
+		d.eei.AddReturnMessage("invalid argument, wanted an address")
+	}
+
+	buff := d.eei.GetStorage(args.CallerAddr)
+	if len(buff) == 0 {
+		d.eei.AddReturnMessage("no sc address under selected user")
 		return vmcommon.UserError
 	}
-	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.ValidatorToDelegation)
-	if err != nil {
-		d.eei.AddReturnMessage(err.Error())
-		return vmcommon.OutOfGas
+
+	found := false
+	for i := 0; i < len(buff); i += lenAddress {
+		savedAddress := buff[i : i+lenAddress]
```

Comment:
Here i+lenAddress could exceed the size of buff for the last iteration if len(buff) will not be a multiplier of lenAddress for some reasons. Should be check through a unit test if it will panic if "i+lenAddress" will exceed the size of the buff. If yes, then we should check that i + lenAddress <= len(buff) and call continue if not

Reply:
this cannot happen. We are always adding a new address into the save []byte array.

Reply:
`	currentStorage := d.eei.GetStorage(args.CallerAddr)
	currentStorage = append(currentStorage, newAddress...)

	d.eei.SetStorage(args.CallerAddr, currentStorage)`

Reply:
ok